reactioncommerce/reaction
# #4441: Improve animations and dynamically import animation libraries to reduce bundle size

Resolves #4441 
Impact: **minor**  
Type: **performance**

## Issue
We currently have `velocity-animate` in the client bundle at 94.5kb. We use Velocity for 5 things:
1) Animate the cart drawer
2) Flash a green color when input fields change
3) Animate the sliding out of the admin panel
4) Animate sliding up & down of admin panel cards
5) Animating the opacity of the overlay when the orders panel is opened


## Solution
Most of the animations are replaceable with CSS. Remove `velocity-animate` and replace those, using a new component called `CSSTransitionGroup` and CSS. Animating the height of admin panel cards requires JS. Use the small `react-animate-height` library for that.

Whether using CSSTransitionGroup or Animate height, the bundle size should not be affected (dynamic imports).

Note: This PR removes `velocity-animate` and `velocity-react` from package.json, and adds `react-transition-group` and `react-animate-height` - both of which are much smaller then Velocity.

## Breaking changes
None

## Testing
1. Visit the homepage and open the cart drawer - confirm improved animation
2. Click a tag in the nav and confirm cart drawer is still open
3. Confirm on tag pages you can toggle cart drawer
4. Change an existing tag's title in the header, confirm that leaving the input after change, or pressing enter after change, the input is flashed green
5. PDP -> Details - Edit a metafield's key, confirm it flashes green. Confirm same with value
6. Edit the product's title inline, and confirm it flashes green leaving the input after change, or pressing enter after change
7. Edit subtitle inline and confirm it flashes green
8. Edit vendor field inline and confirm it flashes green
9. Edit a field in the variant form (other than tax code). Confirm it flashes a lighter green color
10. Confirm admin panel can slide open/closed
11. Confirm admin panel cards can slide up/down
12. Click "Orders" and confirm black overlay fades in
13. Click through to order detail, confirm detail panel slides in
14. Confirm 10-13 while using an RTL language
15. Confirm switching from LTR to RTL language, the admin bar is positioned correctly (no empty scrollbar on RTL, no cut-off icons when switching back to LTR)
16. Confirm double-clicking a product in grid - in edit mode - brings you to PDP without crashing the app or any errors in console.
17. Confirm none of these packages are in the client bundle: `velocity-animate`, `velocity-react`, `react-animate-height`, `react-transition-group`. Run with `export TOOL_NODE_FLAGS='--max-old-space-size=4096' && METEOR_DISABLE_OPTIMISTIC_CACHING=1 meteor run --production --extra-packages bundle-visualizer`

## Before
![velocity-before](https://user-images.githubusercontent.com/849570/43795664-0ce19a34-9a50-11e8-8db3-3dbff390c991.jpg)

## After
![velocity-after](https://user-images.githubusercontent.com/849570/43966359-ae8920fa-9c8f-11e8-87ea-53d0742133c3.jpg)
